wrappers/golang/curves/bls12377/poseidon/poseidon.go

```diff
@@ -3,55 +3,110 @@ package poseidon
 // #cgo CFLAGS: -I./include/
 // #include "poseidon.h"
 import "C"
-
 import (
+	"runtime"
 	"unsafe"
 
 	"github.com/ingonyama-zk/icicle/v2/wrappers/golang/core"
 	cr "github.com/ingonyama-zk/icicle/v2/wrappers/golang/cuda_runtime"
+	bls12_377 "github.com/ingonyama-zk/icicle/v2/wrappers/golang/curves/bls12377"
 )
 
-func GetDefaultPoseidonConfig() core.PoseidonConfig {
-	return core.GetDefaultPoseidonConfig()
+type PoseidonHandler = C.struct_PoseidonInst
+type Poseidon struct {
+	width  uint32
+	handle *PoseidonHandler
 }
 
-func PoseidonHash[T any](scalars, results core.HostOrDeviceSlice, numberOfStates int, cfg *core.PoseidonConfig, constants *core.PoseidonConstants[T]) core.IcicleError {
-	scalarsPointer, resultsPointer, cfgPointer := core.PoseidonCheck(scalars, results, cfg, constants, numberOfStates)
+func Create(arity uint32, alpha uint32, fullRoundsHalf uint32, partialRounds uint32, scalars core.HostOrDeviceSlice, mdsMatrix core.HostOrDeviceSlice, nonSparseMatrix core.HostOrDeviceSlice, sparseMatrices core.HostOrDeviceSlice, domainTag bls12_377.ScalarField, ctx *cr.DeviceContext) (*Poseidon, core.IcicleError) {
+	var poseidon *PoseidonHandler
+	cArity := (C.uint)(arity)
+	cAlpha := (C.uint)(alpha)
+	cFullRoundsHalf := (C.uint)(fullRoundsHalf)
+	cPartialRounds := (C.uint)(partialRounds)
+	cScalars := (*C.scalar_t)(scalars.AsUnsafePointer())
+	cMdsMatrix := (*C.scalar_t)(mdsMatrix.AsUnsafePointer())
+	cNonSparseMatrix := (*C.scalar_t)(nonSparseMatrix.AsUnsafePointer())
+	cSparseMatrices := (*C.scalar_t)(sparseMatrices.AsUnsafePointer())
+	cDomainTag := (*C.scalar_t)(unsafe.Pointer(&domainTag))
+	cCtx := (*C.DeviceContext)(unsafe.Pointer(ctx))
+	__ret := C.bls12_377_poseidon_create_cuda(&poseidon, cArity, cAlpha, cFullRoundsHalf, cPartialRounds, cScalars, cMdsMatrix, cNonSparseMatrix, cSparseMatrices, cDomainTag, cCtx)
+	err := core.FromCudaError((cr.CudaError)(__ret))
+	if err.IcicleErrorCode != core.IcicleSuccess {
+		return nil, err
+	}
+	p := Poseidon{handle: poseidon, width: arity + 1}
+	runtime.SetFinalizer(ctx, p.Delete)
```

Comment:
Why is the Finalizer being set on the `ctx`? Its possible that the context may outlive the need of Poseidon and be used with other primitives making preventing Poseidon from being deleted

Reply:
I guess a simple solution here would be to just store device id inside Poseidon struct

Reply:
Yes, this is what I discussed with @nonam3e:

"Maybe we need to add a DeviceId to the Poseidon struct in Go so that we can call SetFinalizer(&p, deletePoseidon) where deletePoseidon is a wrapper function around p.Delete that creates a new ctx object and assigns the correct DeviceId based on p.DeviceId"

specifically the wrapper function is required because the second argument to `SetFinalizer` must be a function that takes a single parameter of the type of the object in the first argument to `SetFinalizer`